City-of-Helsinki/drupal-helfi-kasvatus-koulutus
# UHF-6317: Add unit categories and separate unit roles

# [UHF-6317](https://helsinkisolutionoffice.atlassian.net/browse/UHF-6317)

## What was done
* Added new `field_categories` field for TPR units.
* During the migration, `field_categories` is set using source data ontologyword IDs.
  * The categories field hold values such as `daycare` to inform that the TPR unit is, in fact, a daycare unit.
* Added three new permissions and roles that use those permissions.
  * The new permissions are used to limit a user to only edit specific type of TPR units.
* Also renamed the school editor label.

## How to install

* Make sure your instance is up and running on correct branch.
  * `git fetch`
  * `git co UHF-6317-separate-tpr-unit-roles`
  *  `make shell` and
      * `composer require drupal/helfi_tpr:dev-UHF-6317-permissions-by-ontology`
  * `make fresh`
* Run `make drush-cr`

## How to test
* Update TPR units by running the migration.
  * `drush shell` and
      * `drush migrate:import tpr_unit`
* You can now see a list of the Ontologyword IDs at the bottom of the unit's edit form.
* You can also see a list of categories (e.g. daycare) if the unit is daycare, comprehensive school or playground.
* Create three new users and give one of the new permissions for each: Daycare editor, Comprehensive school editor, Playground editor.
* Log-in with the new users and check that they can only edit the equivalent TPR units, e.g. the user with Daycare editor role can edit daycare unit pages.
* Check that the new users have reasonable permissions so that they can add and remove media, etc.
* Check that also the translated units work correctly.

Note: Tests are failing until there is a new helfi_tpr release. However, you can run the new tests:
* `make shell` and
  * `vendor/bin/phpunit -c /app/phpunit.xml.dist /app/public/modules/custom`

Also check the code of the linked PR.

## Designers review



## Other PRs
* https://github.com/City-of-Helsinki/drupal-module-helfi-tpr/pull/127
